### PR TITLE
Include post select field

### PR DESCRIPTION
**Pull request by mattheu**
https://github.com/fusioneng/Shortcake/issues/123

This is pretty solid - we've been using it on Fusion.

My one concern is that we're going to have to bundle Select2 into shortcake. 
- Do we have a problem with this?
- Which version should we use? Currently on `3.*` so could be updated to 4.0.
- Can we avoid conflicts with other plugins using it?
- Is it still on the cards to include this in core?

### CODE REVIEW

File: inc/fields/class-field-post-select.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
+<?php
+
+class Shortcake_Field_Post_Select {
```
Comment (danielbachhuber) [inline inc/fields/class-field-post-select.php]: This should technically be `Shortcode_UI`. All of our code-level references should be `shortcode-ui` for consistency.

File: inc/fields/class-field-post-select.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+			self::$instance->setup_actions();
+		}
+		return self::$instance;
+	}
+
+	private function setup_actions() {
+
+		add_filter( 'shortcode_ui_fields', array( $this, 'filter_shortcake_fields' ) );
+
+		add_action( 'admin_enqueue_scripts', array( $this, 'action_admin_enqueue_scripts' ), 100 );
```
Comment (danielbachhuber) [inline inc/fields/class-field-post-select.php]: Needs to respect #280

File: inc/fields/class-field-post-select.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		$nonce               = isset( $_GET['nonce'] ) ? sanitize_text_field( $_GET['nonce'] ) : null;
+		$requested_shortcode = isset( $_GET['shortcode'] ) ? sanitize_text_field( $_GET['shortcode'] ) : null;
+		$requested_attr      = isset( $_GET['attr'] ) ? sanitize_text_field( $_GET['attr'] ) : null;
+		$response            = array( 'posts' => array(), 'found_posts'=> 0, 'posts_per_page' => 0 );
+
+		$shortcodes = Shortcode_UI::get_instance()->get_shortcodes();
+
+		if ( ! wp_verify_nonce( $nonce, 'shortcake_field_post_select' ) ) {
+			wp_send_json_error( $response );
+		}
```
Comment (danielbachhuber) [inline inc/fields/class-field-post-select.php:113]: Do we need a capability check in here too?

File: js/field-post-select.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		}
+
+	} );
+
+	/**
+	 * Extending SUI Media Controller to hide Select2 UI Drop-Down when menu
+	 * changes in Meida modal
+	 * 1. going back/forth between different shortcakes (refresh)
+	 * 2. changing the menu in left column (deactivate)
+	 * 3. @TODO closing the modal.
```
Comment (danielbachhuber) [inline js/field-post-select.js:167]: Do we need to address this?

File: shortcode-ui.php [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?php
 /**
  * Plugin Name: Shortcode UI
- * Version: v0.3-alpha
+ * Version: v0.2.0-dev
```
Comment (danielbachhuber) [inline shortcode-ui.php]: Can you fix this change?

**Comment by danielbachhuber** [comment]
> Do we have a problem with this?

Seems like its utility outweighs the cost. Should it be abstracted to use other types too, like Users and Terms?

> Which version should we use? Currently on 3.\* so could be updated to 4.0.

Not sure of the nuances of one vs another?

> Can we avoid conflicts with other plugins using it?

Not without namespacing it, me thinks. We should enqueue as `select2` and respect when it's already been loaded.

**Comment by mattheu** [comment]
V3 and V4 are not compatable with each other. So lets keep it at version 3. Its done and other people are more likeley to be using this version too. (Including us ;))

**Comment by danielbachhuber** [comment]
> V3 and V4 are not compatable with each other. So lets keep it at version 3. Its done and other people are more likeley to be using this version too. (Including us ;))

Sounds good.

**Comment by danielbachhuber** [comment]
Let me know when this is ready to :ship:

**Comment by mattheu** [comment]
I've removed the code that attempts to destroy the field when you switch or close whilst the field is still open. It just doesn't seem to work.  However I don't think this is neccessarily a blocker - still a useful field and we can spin this off into another ticket

@jitendraharpalani58  I think you wrote this originally? Maybe you can shed some light on it?

**Comment by danielbachhuber** [comment]
> I've removed the code that attempts to destroy the field when you switch or close whilst the field is still open. It just doesn't seem to work

I've reverted. It works in the insert context, but not the edit context.

File: inc/fields/class-field-post-select.php (hunk 2 above)
Comment (danielbachhuber) [inline inc/fields/class-field-post-select.php]: 1146f8a03d488b8b1988e815b9308b79879e1861

File: inc/fields/class-field-post-select.php (hunk 3 above)
Comment (danielbachhuber) [inline inc/fields/class-field-post-select.php:113]: It looks like `wp_ajax_find_posts()` doesn't have a capability check, so we're good.

**Pull request merged**